assets/js/modules/analytics-4/utils/withCustomDimensions.js

```diff
@@ -0,0 +1,374 @@
+/**
+ * `withCustomDimensions` HOC.
+ *
+ * Site Kit by Google, Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import {
+	createInterpolateElement,
+	useCallback,
+	useEffect,
+	useMemo,
+} from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { Button } from 'googlesitekit-components';
+import Data from 'googlesitekit-data';
+import { CORE_FORMS } from '../../../googlesitekit/datastore/forms/constants';
+import { CORE_LOCATION } from '../../../googlesitekit/datastore/location/constants';
+import { CORE_SITE } from '../../../googlesitekit/datastore/site/constants';
+import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
+import { EDIT_SCOPE as ANALYTICS_EDIT_SCOPE } from '../../analytics/datastore/constants';
+import {
+	FORM_CUSTOM_DIMENSIONS_CREATE,
+	MODULES_ANALYTICS_4,
+} from '../datastore/constants';
+import { KEY_METRICS_WIDGETS } from '../../../components/KeyMetrics/key-metrics-widgets';
+import Link from '../../../components/Link';
+import MetricTileError from '../../../components/KeyMetrics/MetricTileError';
+import MetricTileWrapper from '../../../components/KeyMetrics/MetricTileWrapper';
+import {
+	ERROR_CODE_MISSING_REQUIRED_SCOPE,
+	ERROR_REASON_BAD_REQUEST,
+	isInsufficientPermissionsError,
+} from '../../../util/errors';
+const { useSelect, useDispatch } = Data;
+
+export default function withCustomDimensions( options = {} ) {
+	const { dimensions, infoTooltip, reportOptions, title } = options;
```

Comment:
How are we going to pass report options into this HOC? Usually all our report options are dynamic and depend on the selected date range which means that we can't just pass an object with parameters into this HOC.